### PR TITLE
few memory leaks fixed

### PR DESCRIPTION
**Pull request by NaveenBoggarapu**
Few memory leaks observed are fixed.

### CODE REVIEW

File: libr/anal/anal.c [hunk 1]
```diff
@@ -118,17 +118,15 @@ R_API RAnal *r_anal_new() {
 	anal->plugins = r_list_newf ((RListFree) r_anal_plugin_free);
 	if (anal->plugins) {
 		for (i=0; anal_static_plugins[i]; i++) {
-			static_plugin = R_NEW (RAnalPlugin);
-			*static_plugin = *anal_static_plugins[i];
-			r_anal_add (anal, static_plugin);
+			r_anal_add (anal, anal_static_plugins[i]);
```
Comment (radare) [inline libr/anal/anal.c]: this is wrong. because r_anal_free will try to liberate them and crash

File: libr/anal/anal.c [hunk 2]
```diff
@@ -118,17 +118,15 @@ R_API RAnal *r_anal_new() {
...
 		}
 	}
 	return anal;
 }
 
 R_API void r_anal_plugin_free (RAnalPlugin *p) {
 	if (p && p->fini) {
-		p->fini (p);
+		p->fini (NULL);
 	}
```
Comment (radare) [inline libr/anal/anal.c]: then plugins cannot be properly finalized

File: libr/asm/asm.c [hunk 3]
```diff
@@ -139,11 +139,7 @@ R_API RAsm *r_asm_new() {
 		return NULL;
 	}
 	for (i=0; asm_static_plugins[i]; i++) {
-		static_plugin = R_NEW (RAsmPlugin);
-		if (!static_plugin) continue;
-		// memleak here
-		memcpy (static_plugin, asm_static_plugins[i], sizeof (RAsmPlugin));
-		r_asm_add (a, static_plugin);
+		r_asm_add (a, asm_static_plugins[i]);
```
Comment (radare) [inline libr/asm/asm.c]: same as above

File: libr/bin/bin.c [hunk 4]
```diff
@@ -302,7 +302,7 @@ static RList *get_strings(RBinFile *a, int min, int dump) {
 		/* dump to stdout, not stored in list */
 		ret = NULL;
 	} else {
-		ret = r_list_newf (free);
+		ret = r_list_newf (r_bin_string_free);
```
Comment (radare) [inline libr/bin/bin.c]: lgtm

File: libr/debug/plugin.c [hunk 5]
```diff
@@ -36,8 +36,10 @@ R_API bool r_debug_use(RDebug *dbg, const char *str) {
 		char *p = dbg->h->reg_profile (dbg);
 		if (p) {
 			r_reg_set_profile_string (dbg->reg, p);
-			if (dbg->anal)
+			if (dbg->anal){
```
Comment (radare) [inline libr/debug/plugin.c]: missing space after parenthesis

File: libr/debug/plugin.c [hunk 6]
```diff
@@ -36,8 +36,10 @@ R_API bool r_debug_use(RDebug *dbg, const char *str) {
...
+				if (dbg->anal->reg) r_reg_free (dbg->anal->reg);
```
Comment (radare) [inline libr/debug/plugin.c]: unnecessary if

File: libr/parse/parse.c [hunk 7]
```diff
@@ -22,10 +22,7 @@ R_API RParse *r_parse_new() {
 	p->notin_flagspace = -1;
 	p->flagspace = -1;
 	for (i=0; parse_static_plugins[i]; i++) {
-		static_plugin = R_NEW (RParsePlugin);
-		memcpy (static_plugin, parse_static_plugins[i],
-			sizeof (RParsePlugin));
-		r_parse_add (p, static_plugin);
+		r_parse_add (p, parse_static_plugins[i]);
```
Comment (radare) [inline libr/parse/parse.c]: same as the other cases

File: libr/reg/reg.c [hunk 8]
```diff
@@ -119,7 +122,7 @@ R_API void r_reg_free(RReg *reg) {
 		return;
 
 	for (i = 0; i < R_REG_TYPE_LAST; i++) {
-		r_list_purge (reg->regset[i].pool);
+		r_list_free (reg->regset[i].pool);
```
Comment (radare) [inline libr/reg/reg.c]: uhm i think there was a reason to use purge instead of free here..cant remember now

File: libr/util/buf.c [hunk 9]
```diff
@@ -557,12 +557,16 @@ R_API void r_buf_deinit(RBuffer *b) {
 	if (b->mmap) {
 		r_file_mmap_free (b->mmap);
 		b->mmap = NULL;
-	} else free (b->buf);
+	} else {
+		free (b->buf);
```
Comment (radare) [inline libr/util/buf.c]: you can use R_FREE to avoid having to free+NULL

**Comment by radare** [comment]
i would like to discuss/review the plugin lifecycle leaks before merging this. can you please update the pr with the changes suggested?

File: libr/anal/anal.c (hunk 1 above)
Comment (NaveenBoggarapu) [inline libr/anal/anal.c]: No, it will not.
When r_anal_free

File: libr/anal/anal.c (hunk 2 above)
Comment (NaveenBoggarapu) [inline libr/anal/anal.c]: None of the plugins apart from `r_anal_plugin_tms320` have `fini` in the declaration, or they are updated during execution.

And in `r_anal_plugin_tms320` `fini` doesn't use passed pointer for finalizing.

File: libr/asm/asm.c (hunk 3 above)
Comment (NaveenBoggarapu) [inline libr/asm/asm.c]: Same as above, most of these plugins have their `fini` as `NULL`, and the one's which have `fini` doesn't use the passed pointer.

**Comment by NaveenBoggarapu** [comment]
I have updated the PR.
and regarding the plugin lifecycle, is my understanding wrong?

File: libr/reg/reg.c (hunk 8 above)
Comment (NaveenBoggarapu) [inline libr/reg/reg.c]: My observation is right after this call, `reg->regset[i].pool` is set to `NULL` making no one pointing to the memory locations, so I thought of using free instead of purge.

I think the reason for using purge instead of free is in `r_reg_arena_push` function
 `r_list_push (reg->regset[i].pool, b);`
 is used to push arena in to pool. So purge was used instead of free since there is a pointer in the list, which was pointed using `reg->regset[i].arena` but since we are any way not freeing any arena while freeing `reg` I think it should be valid.

**Comment by radare** [comment]
Rebase and checkout the output of travis, because after your commit it's full of crashes and XX. mainly in java.

https://api.travis-ci.org/jobs/122725957/log.txt?deansi=true

**Comment by NaveenBoggarapu** [comment]
Ok, I will check that.

File: libr/anal/anal.c (hunk 2 above)
Comment (alvarofe) [inline libr/anal/anal.c]: if there is no ini in the declaration is NULL thus is not called but you can't pass NULL to fini because if i develop an anal plugin with a fini function will not be called as i expect. 

Regarding plugin_tms320 it doesn't use it because it has an static struct so the parameter is unused but the API is the way it is.

Comment (NaveenBoggarapu) [inline libr/anal/anal.c]: then it should be the same with `asm_static_plugins`, but `plugin_free` in asm.c doesn't pass pointer to fini.
It might need a change accordingly. 

There should also be a common method for static_plugins, a way to free the memory allocated to them.

**Comment by NaveenBoggarapu** [comment]
travis is passing now

**Comment by radare** [comment]
Looks good now. thanks!

**Comment by alvarofe** [comment]
Revert this because i bet are gonna be problems with use after free, this need to be tested thoroughly, the fini is passing NULL and such. If iirc i did this some time ago but there were a lot of use after free. I will try to review this weekend with time and compile it with ASAN.

**Comment by radare** [comment]
We have time for testing now. Still 6 weeks for release. Lets do craZy shit now
